voxblox/include/voxblox/simulation/objects.h

```diff
@@ -242,6 +242,147 @@ class Plane : public Object {
   Point normal_;
 };
 
+class Cylinder : public Object {
+ public:
+  Cylinder(const Point& center, FloatingPoint radius, FloatingPoint height)
+      : Object(center, Type::kCylinder), radius_(radius), height_(height) {}
+  Cylinder(const Point& center, FloatingPoint radius, FloatingPoint height,
+           const Color& color)
+      : Object(center, Type::kCylinder, color),
+        radius_(radius),
+        height_(height) {}
+
+  virtual FloatingPoint getDistanceToPoint(const Point& point) const {
+    // From: https://math.stackexchange.com/questions/2064745/
+    // 3 cases, depending on z of point.
+    // First case: in plane with the cylinder. This also takes care of inside
+    // case.
+    FloatingPoint distance = 0.0;
+
+    FloatingPoint min_z_limit = center_.z() - height_ / 2.0;
+    FloatingPoint max_z_limit = center_.z() + height_ / 2.0;
+    if (point.z() >= min_z_limit && point.z() <= max_z_limit) {
+      distance = (point.head<2>() - center_.head<2>()).norm() - radius_;
+    } else if (point.z() > max_z_limit) {
+      // Case 2: above the cylinder.
+      distance = std::sqrt(
+          std::max((point.head<2>() - center_.head<2>()).squaredNorm() -
+                       radius_ * radius_,
+                   0.0) +
+          (point.z() - max_z_limit) * (point.z() - max_z_limit));
+    } else {
+      // Case 3: below cylinder.
+      distance = std::sqrt(
+          std::max((point.head<2>() - center_.head<2>()).squaredNorm() -
+                       radius_ * radius_,
+                   0.0) +
+          (point.z() - min_z_limit) * (point.z() - min_z_limit));
+    }
+    return distance;
+  }
+
+  virtual bool getRayIntersection(const Point& ray_origin,
+                                  const Point& ray_direction,
+                                  FloatingPoint max_dist,
+                                  Point* intersect_point,
+                                  FloatingPoint* intersect_dist) const {
+    // From http://woo4.me/wootracer/cylinder-intersection/
+    // and http://www.cl.cam.ac.uk/teaching/1999/AGraphHCI/SMAG/node2.html
+    // Define ray as P = E + tD, where E is ray_origin and D is ray_direction.
+    // We define our cylinder as centered in the xy coordinate system, so
+    // E in this case is actually ray_origin - center_.
+    Point vector_E = ray_origin - center_;
+    Point vector_D = ray_direction;  // Axis aligned.
+
+    FloatingPoint a = vector_D.x() * vector_D.x() + vector_D.y() * vector_D.y();
+    FloatingPoint b =
+        2 * vector_E.x() * vector_D.x() + 2 * vector_E.y() * vector_D.y();
+    FloatingPoint c = vector_E.x() * vector_E.x() +
+                      vector_E.y() * vector_E.y() - radius_ * radius_;
+
+    // t = (-b +- sqrt(b^2 - 4ac))/2a
+    // t only has solutions if b^2 - 4ac >= 0
+    FloatingPoint t1 = -1.0;
+    FloatingPoint t2 = -1.0;
+
+    // Make sure we don't divide by 0.
+    if (std::abs(a) < 1e-6) {
```

Comment:
Should we just define a global kEpsilon for voxblox it seems to be needed in a few places